e2e_dra: tests panic due to too short TContext CleanupGracePeriod

TContext hardcodes [CleanupGracePeriod = 5 sec](https://github.com/kubernetes/kubernetes/blob/v1.36.0/test/utils/ktesting/tcontext.go#L60), which is not enough for e2e_dra tests to shutdown properly. This value should be configurable.

Here is a beginning of panic (full traceback is too long) caused by this issue taken from [this test run](https://storage.googleapis.com/kubernetes-ci-logs/pr-logs/pull/138791/pull-kubernetes-dra-integration/2052471685096935424/build-log.txt):

```
    client.go:193: I0507 20:23:57.143582] Waiting for ResourceSlices of driver defaultpartitionable-devices.k8s.io to be removed...
    contexthelper.go:75: 
        INFO: canceling context: test suite deadline (2026-05-07 20:24:25 +0000 UTC) is close, need to clean up before the 5s cleanup grace period
        
panic: test timed out after 30m0s
	running tests:
		TestUpgradeDowngrade (9m54s)

goroutine 4776 [running]:
testing.(*M).startAlarm.func1()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/testing/testing.go:2802 +0x605
created by time.goFunc
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/time/sleep.go:215 +0x45

goroutine 1 [chan receive, 10 minutes]:
testing.(*T).Run(0xc000366248, {0x59340b7, 0x14}, 0x5c63398)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/testing/testing.go:2109 +0xb3e
testing.runTests.func1(0xc000366248)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/testing/testing.go:2585 +0x85
testing.tRunner(0xc000366248, 0xc00008bad0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/testing/testing.go:2036 +0x21d
testing.runTests({0x5920c3c, 0x11}, {0x595f15d, 0x1e}, 0xc00078e0a8, {0x7e42840, 0x2, 0x2}, {0xc2775b1e44154785, 0x1a349b551b1, ...})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/testing/testing.go:2583 +0x9ea
testing.(*M).Run(0xc0006a6a00)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/testing/testing.go:2443 +0xf4c
main.main()
	_testmain.go:48 +0x165

goroutine 2391 [select]:
golang.org/x/time/rate.(*Limiter).wait(0xc000fd4280, {0x5cae410, 0xc0012a06c0}, 0x1, {0x5cdacb8?, 0x0?, 0x7fe6ec0?}, 0x5c65c20)
	/home/prow/go/src/k8s.io/kubernetes/vendor/golang.org/x/time/rate/rate.go:286 +0x50f
golang.org/x/time/rate.(*Limiter).WaitN(0xc000fd4280, {0x5cae410, 0xc0012a06c0}, 0x1)
	/home/prow/go/src/k8s.io/kubernetes/vendor/golang.org/x/time/rate/rate.go:249 +0x5a
golang.org/x/time/rate.(*Limiter).Wait(...)
	/home/prow/go/src/k8s.io/kubernetes/vendor/golang.org/x/time/rate/rate.go:234
k8s.io/client-go/util/flowcontrol.(*tokenBucketRateLimiter).Wait(0xc00077cb70, {0x5cae410, 0xc0012a06c0})
	/home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/util/flowcontrol/throttle.go:131 +0x4f
k8s.io/client-go/rest.(*Request).tryThrottleWithInfo(0xc000f41800, {0x5cae410, 0xc0012a06c0}, {0x0, 0x0})
	/home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/rest/request.go:673 +0xd9
k8s.io/client-go/rest.(*Request).tryThrottle(...)
	/home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/rest/request.go:697
k8s.io/client-go/rest.(*Request).request(0xc000f41800, {0x5cae410, 0xc0012a06c0}, 0xc00135eed8)
	/home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/rest/request.go:1054 +0x3f5
k8s.io/client-go/rest.(*Request).Do(0xc000f41800, {0x5cae410, 0xc0012a06c0})
	/home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/rest/request.go:1136 +0x1f1
k8s.io/client-go/gentype.(*alsoLister[...]).List(0x5c89580, {0x5cae410, 0xc0012a06c0}, {{{0x0, 0x0}, {0x0, 0x0}}, {0x0, 0x0}, {0xc00103c6f0, ...}, ...})
	/home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/client-go/gentype/type.go:183 +0x451
k8s.io/dynamic-resource-allocation/client.(*convertingClient[...]).List.func1()
	/home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/dynamic-resource-allocation/client/generic.go:197 +0x2eb
k8s.io/dynamic-resource-allocation/client.(*callSequence[...]).run(0x51828e0)
	/home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/dynamic-resource-allocation/client/generic.go:251 +0xbb
k8s.io/dynamic-resource-allocation/client.(*convertingClient[...]).List(0x5d02000, {0x5cae410, 0xc0012a06c0}, {{{0x0, 0x0}, {0x0, 0x0}}, {0x0, 0x0}, {0xc00103c6f0, ...}, ...})
	/home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/dynamic-resource-allocation/client/generic.go:200 +0x145
k8s.io/kubernetes/test/e2e/framework.ListObjects[...].func1()
	/home/prow/go/src/k8s.io/kubernetes/test/e2e/framework/get.go:61 +0x174
k8s.io/kubernetes/test/e2e/framework.ListObjects[...].func2()
	/home/prow/go/src/k8s.io/kubernetes/test/e2e/framework/get.go:72 +0x7e
k8s.io/kubernetes/test/e2e/dra/utils.(*Driver).IsGone.(*Driver).NewGetSlices.func1({{{0x5cae528, 0xc00125fa90}, {{0x5cdacb8, 0xc000366908}}, 0x5c64ea0, 0xc00090c3a0, {0x0, 0x0}, 0x0, {0x58e961c, ...}, ...}})
	/home/prow/go/src/k8s.io/kubernetes/test/e2e/dra/utils/deploy.go:336 +0x25f
reflect.Value.call({0x52a1940?, 0xc000f8a530?, 0xc001360260?}, {0x58e3338, 0x4}, {0xc0026e4198, 0x1, 0x178541e?})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/reflect/value.go:586 +0x1232
reflect.Value.Call({0x52a1940?, 0xc000f8a530?, 0x5cdacb8?}, {0xc0026e4198, 0x1, 0x1})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/reflect/value.go:369 +0xb6
k8s.io/kubernetes/test/utils/client-go/ktesting.TContext.newAsyncAssertion.func1({0xc0026e4198, 0x1, 0x1})
	/home/prow/go/src/k8s.io/kubernetes/test/utils/client-go/ktesting/assert.go:156 +0x45a
reflect.Value.call({0x5370b40?, 0xc000b3eb40?, 0x1697938?}, {0x58e3338, 0x4}, {0xc001403ed8, 0x1, 0xc000600808?})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/reflect/value.go:586 +0x1232
reflect.Value.Call({0x5370b40?, 0xc000b3eb40?, 0xc00136169a?}, {0xc001403ed8, 0x1, 0x1})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/reflect/value.go:369 +0xb6
github.com/onsi/gomega/internal.(*AsyncAssertion).buildActualPoller.func3()
	/home/prow/go/src/k8s.io/kubernetes/vendor/github.com/onsi/gomega/internal/async_assertion.go:337 +0x23b
github.com/onsi/gomega/internal.(*AsyncAssertion).match(0xc000435420, {0x5ca3150, 0xc000a7c0c0}, 0x1, {0x0, 0x0, 0x0})
	/home/prow/go/src/k8s.io/kubernetes/vendor/github.com/onsi/gomega/internal/async_assertion.go:560 +0xf6c
github.com/onsi/gomega/internal.(*AsyncAssertion).Should(0xc000435420, {0x5ca3150, 0xc000a7c0c0}, {0x0, 0x0, 0x0})
	/home/prow/go/src/k8s.io/kubernetes/vendor/github.com/onsi/gomega/internal/async_assertion.go:145 +0xc5
k8s.io/kubernetes/test/e2e/dra/utils.(*Driver).IsGone(0xc000ff7180, {{{0x5cae798, 0xc000a7c020}, {{0x5cdacb8, 0xc000366908}}, 0x5c64ea0, 0xc001034b00, {0x0, 0x0}, 0x0, ...}})
	/home/prow/go/src/k8s.io/kubernetes/test/e2e/dra/utils/deploy.go:1053 +0x2bb
k8s.io/kubernetes/test/utils/client-go/ktesting.TContext.CleanupCtx.func1({{0x5cae798, 0xc000a7c020}, {{0x5cdacb8, 0xc000366908}}, 0x5c64ea0, 0xc001034b00, {0x0, 0x0}, 0x0, {0x58e961c, ...}, ...})
	/home/prow/go/src/k8s.io/kubernetes/test/utils/client-go/ktesting/client.go:193 +0xe2
k8s.io/kubernetes/test/utils/ktesting.TContext.CleanupCtx.func2()
	/home/prow/go/src/k8s.io/kubernetes/test/utils/ktesting/tcontext.go:497 +0x205
testing.(*common).Cleanup.func1()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/testing/testing.go:1317 +0x169
testing.(*common).runCleanup(0xc000366908, 0x0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/testing/testing.go:1667 +0x226
testing.tRunner.func2()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/testing/testing.go:2030 +0x4d
testing.tRunner(0xc000366908, 0x5c63398)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/testing/testing.go:2042 +0x256
created by testing.(*T).Run in goroutine 1
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/cache/mod/golang.org/toolchain@v0.0.1-go1.26.3.linux-amd64/src/testing/testing.go:2101 +0xb13
...
```

It would be nice to add `CleanupGracePeriod` to the TContext object and provide an option to set it from the caller side.

/cc @pohly 
/sig testing

--- a/test/utils/ktesting/initoption/initoption.go
+++ b/test/utils/ktesting/initoption/initoption.go
@@ -16,7 +16,11 @@ limitations under the License.
 
 package initoption
 
-import "k8s.io/kubernetes/test/utils/ktesting/internal"
+import (
+	"time"
+
+	"k8s.io/kubernetes/test/utils/ktesting/internal"
+)
 
 // InitOption is a functional option for Init and InitCtx.
 type InitOption func(c *internal.InitConfig)
@@ -36,5 +40,20 @@ func PerTestOutput(enabled bool) InitOption {
 func BufferLogs(enabled bool) InitOption {
 	return func(c *internal.InitConfig) {
 		c.BufferLogs = enabled
+	}
+}
+
+// WithCleanupGracePeriod overrides the default cleanup grace period used by
+// [Init]. The cleanup grace period is the time reserved before the test-suite
+// deadline so that cleanup callbacks can complete before the test binary is
+// killed.
+//
+// A non-positive value is ignored and the default is used instead.
+//
+// When using Ginkgo to manage the test suite this option has no effect because
+// Ginkgo itself manages timeouts.
+func WithCleanupGracePeriod(d time.Duration) InitOption {
+	return func(c *internal.InitConfig) {
+		c.CleanupGracePeriod = d
 	}
 }

--- a/test/utils/ktesting/internal/config.go
+++ b/test/utils/ktesting/internal/config.go
@@ -16,7 +16,10 @@ limitations under the License.
 
 package internal
 
+import "time"
+
 type InitConfig struct {
-	PerTestOutput bool
-	BufferLogs    bool
+	PerTestOutput      bool
+	BufferLogs         bool
+	CleanupGracePeriod time.Duration // Zero means use the package default (DefaultCleanupGracePeriod).
 }

--- a/test/utils/ktesting/tcontext.go
+++ b/test/utils/ktesting/tcontext.go
@@ -40,18 +40,20 @@ import (
 // used to capture log output in memory to check it in tests.
 type Underlier = ktesting.Underlier
 
-// CleanupGracePeriod is the time that a [TContext] gets canceled before the
-// deadline of its underlying test suite (usually determined via "go test
+// DefaultCleanupGracePeriod is the time that a [TContext] gets canceled before
+// the deadline of its underlying test suite (usually determined via "go test
 // -timeout"). This gives the running test(s) time to fail with an informative
 // timeout error. After that, all cleanup callbacks then have the remaining
 // time to complete before the test binary is killed.
 //
-// For this to work, each blocking calls in a test must respect the
+// For this to work, each blocking call in a test must respect the
 // cancellation of the [TContext].
 //
 // When using Ginkgo to manage the test suite and running tests, the
-// CleanupGracePeriod is ignored because Ginkgo itself manages timeouts.
-const CleanupGracePeriod = 5 * time.Second
+// cleanup grace period is ignored because Ginkgo itself manages timeouts.
+//
+// This value can be overridden per-[TContext] via [initoption.WithCleanupGracePeriod].
+const DefaultCleanupGracePeriod = 5 * time.Second
 
 // TB is the interface common to [testing.T], [testing.B], [testing.F] and
 // [github.com/onsi/ginkgo/v2] which ktesting relies upon itself or
@@ -172,15 +174,23 @@ func Init(tb TB, opts ...InitOption) TContext {
 		header = klogHeader
 	}
 
+	// Resolve the effective cleanup grace period: use the caller-supplied value
+	// if positive, otherwise fall back to DefaultCleanupGracePeriod.
+	gracePeriod := c.CleanupGracePeriod
+	if gracePeriod <= 0 {
+		gracePeriod = DefaultCleanupGracePeriod
+	}
+
 	var cancelTimeout func(cause string)
 	if deadline != nil {
 		timeLeft := time.Until(*deadline)
-		timeLeft -= CleanupGracePeriod
-		ctx, cancelTimeout = withTimeout(ctx, tb, timeLeft, fmt.Sprintf("test suite deadline (%s) is close, need to clean up before the %s cleanup grace period", deadline.Truncate(time.Second), CleanupGracePeriod))
+		timeLeft -= gracePeriod
+		ctx, cancelTimeout = withTimeout(ctx, tb, timeLeft, fmt.Sprintf("test suite deadline (%s) is close, need to clean up before the %s cleanup grace period", deadline.Truncate(time.Second), gracePeriod))
 	}
 
 	// Construct new TContext with context and settings as determined above.
 	tCtx := InitCtx(ctx, tb)
+	tCtx.cleanupGracePeriod = gracePeriod
 	tCtx.isSyncTest = isSyncTest
 	if cancelTimeout != nil {
 		tCtx.cancel = cancelTimeout
@@ -240,12 +250,19 @@ type InitOption = initoption.InitOption
 
 // InitCtx is a variant of [Init] which uses an already existing context and
 // whatever logger and timeouts are stored there.
-// Functional options are part of the API, but currently
-// there are none which have an effect.
-func InitCtx(ctx context.Context, tb TB, _ ...InitOption) TContext {
+func InitCtx(ctx context.Context, tb TB, opts ...InitOption) TContext {
+	c := internal.InitConfig{}
+	for _, opt := range opts {
+		opt(&c)
+	}
+	gracePeriod := c.CleanupGracePeriod
+	if gracePeriod <= 0 {
+		gracePeriod = DefaultCleanupGracePeriod
+	}
 	return TContext{
-		Context:   ctx,
-		testingTB: testingTB{TB: tb},
+		Context:            ctx,
+		testingTB:          testingTB{TB: tb},
+		cleanupGracePeriod: gracePeriod,
 	}
 }
 
@@ -414,6 +431,11 @@ type TContext struct {
 	//
 	// Used by WithError.
 	capture *capture
+
+	// cleanupGracePeriod is the effective cleanup grace period for this context.
+	// It is always non-zero: both Init and InitCtx resolve it from the supplied
+	// options or fall back to DefaultCleanupGracePeriod.
+	cleanupGracePeriod time.Duration
 }
 
 type capture struct {

--- a/test/utils/ktesting/tcontext_internal_test.go
+++ b/test/utils/ktesting/tcontext_internal_test.go
@@ -1,0 +1,99 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file is in package ktesting (not ktesting_test) so that it can access
+// unexported fields of TContext.
+package ktesting
+
+import (
+	"context"
+	"testing"
+	"testing/synctest"
+	"time"
+
+	"github.com/onsi/gomega"
+
+	"k8s.io/kubernetes/test/utils/ktesting/initoption"
+)
+
+// deadlineT2 mirrors the deadlineT helper in ktesting_test.
+type deadlineT2 struct {
+	TB
+	deadline *time.Time
+}
+
+func (t *deadlineT2) Deadline() (time.Time, bool) {
+	if t.deadline == nil {
+		return time.Time{}, false
+	}
+	return *t.deadline, true
+}
+
+// TestDefaultCleanupGracePeriod verifies that cleanupGracePeriod is set to
+// DefaultCleanupGracePeriod when no override is given.
+func TestDefaultCleanupGracePeriod(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		mockDeadline := time.Date(2000, 1, 1, 0, 0, 0, 0, time.UTC)
+		mockT := &deadlineT2{TB: t, deadline: &mockDeadline}
+		tCtx := Init(mockT)
+		if tCtx.cleanupGracePeriod != DefaultCleanupGracePeriod {
+			t.Errorf("expected cleanupGracePeriod %v, got %v", DefaultCleanupGracePeriod, tCtx.cleanupGracePeriod)
+		}
+	})
+}
+
+// TestCustomCleanupGracePeriod verifies that WithCleanupGracePeriod stores the
+// override in cleanupGracePeriod and shifts the effective deadline accordingly.
+func TestCustomCleanupGracePeriod(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		mockDeadline := time.Date(2000, 1, 1, 0, 0, 0, 0, time.UTC)
+		mockT := &deadlineT2{TB: t, deadline: &mockDeadline}
+		custom := 30 * time.Second
+		tCtx := Init(mockT, initoption.WithCleanupGracePeriod(custom))
+
+		if tCtx.cleanupGracePeriod != custom {
+			t.Errorf("expected cleanupGracePeriod %v, got %v", custom, tCtx.cleanupGracePeriod)
+		}
+
+		actualDeadline, ok := tCtx.Deadline()
+		if !ok {
+			t.Fatal("expected a deadline but got none")
+		}
+		expect := mockDeadline.Add(-custom)
+		tCtx.Expect(actualDeadline).To(
+			gomega.BeTemporally("==", expect),
+			"context deadline should be shifted by the custom grace period")
+	})
+}
+
+// TestInitCtxCleanupGracePeriod verifies that InitCtx applies
+// WithCleanupGracePeriod via the previously ignored opts parameter.
+func TestInitCtxCleanupGracePeriod(t *testing.T) {
+	custom := 42 * time.Second
+	tCtx := InitCtx(context.Background(), t, initoption.WithCleanupGracePeriod(custom))
+	if tCtx.cleanupGracePeriod != custom {
+		t.Errorf("expected cleanupGracePeriod %v, got %v", custom, tCtx.cleanupGracePeriod)
+	}
+}
+
+// TestInitCtxDefaultCleanupGracePeriod verifies that InitCtx falls back to
+// DefaultCleanupGracePeriod when no option is given.
+func TestInitCtxDefaultCleanupGracePeriod(t *testing.T) {
+	tCtx := InitCtx(context.Background(), t)
+	if tCtx.cleanupGracePeriod != DefaultCleanupGracePeriod {
+		t.Errorf("expected cleanupGracePeriod %v, got %v", DefaultCleanupGracePeriod, tCtx.cleanupGracePeriod)
+	}
+}

--- a/test/utils/ktesting/tcontext_test.go
+++ b/test/utils/ktesting/tcontext_test.go
@@ -75,7 +75,7 @@ func TestNormalInit(t *testing.T) {
 	// The outcome depends on how the unit test was started.
 	// See below for deterministic deadline/no deadline testing.
 	expectDeadline, expectOK := t.Deadline()
-	expectDeadline = expectDeadline.Add(-ktesting.CleanupGracePeriod)
+	expectDeadline = expectDeadline.Add(-ktesting.DefaultCleanupGracePeriod)
 	tCtx := ktesting.Init(t)
 	actualDeadline, actualOK := tCtx.Deadline()
 	tCtx.Expect(actualOK).To(gomega.Equal(expectOK), "have deadline")
@@ -104,7 +104,7 @@ func TestDeadline(t *testing.T) {
 		tCtx := ktesting.Init(mockT)
 		actualDeadline, ok := tCtx.Deadline()
 		if ok {
-			expectDeadline := mockDeadline.Add(-ktesting.CleanupGracePeriod)
+			expectDeadline := mockDeadline.Add(-ktesting.DefaultCleanupGracePeriod)
 			tCtx.Expect(actualDeadline).To(gomega.BeTemporally("==", expectDeadline), "deadline")
 		} else {
 			tCtx.Error("Expected a deadline, got none")

